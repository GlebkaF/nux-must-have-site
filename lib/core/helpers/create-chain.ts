import { ModulationType } from "../blocks/modulation";
import { CompressorType } from "../blocks/compressor";
import { NoiseGateType } from "../blocks/noisegate";
import { ReverbType } from "../blocks/reverb";
import { Blocks, Chain } from "../interface";
import { EffectType } from "../blocks/effect";
import { AmplifierType } from "../blocks/amplifier";
import { CabinetType } from "../blocks/cabinet";
import { EqType } from "../blocks/eq";
import { DelayType } from "../blocks/delay";
import { config } from "lib/core/config";

export interface BlockTypesConfig {
  [Blocks.Noisegate]: NoiseGateType | null;
  [Blocks.Compressor]: CompressorType | null;
  [Blocks.Modulation]: ModulationType | null;
  [Blocks.Effect]: EffectType | null;
  [Blocks.Amplifier]: AmplifierType | null;
  [Blocks.Cabinet]: CabinetType | null;
  [Blocks.Eq]: EqType | null;
  [Blocks.Reverb]: ReverbType | null;
  [Blocks.Delay]: DelayType | null;
}

export function createDefaultChain(): Chain {
  return createEmptyChain({
    [Blocks.Noisegate]: NoiseGateType.NoiseGate,
    [Blocks.Compressor]: CompressorType.KComp,
    [Blocks.Modulation]: ModulationType.CE1,
    [Blocks.Effect]: EffectType.DistortionPlus,
    [Blocks.Amplifier]: AmplifierType.JazzClean,
    [Blocks.Cabinet]: CabinetType.JZ120,
    [Blocks.Eq]: EqType.SixBand,
    [Blocks.Reverb]: ReverbType.Room,
    [Blocks.Delay]: DelayType.AnalogDelay,
  });
}

export function createEmptyChain(types: BlockTypesConfig): Chain {
  const createTypeParams = <T extends Blocks>(
    block: T,
    type: string | null
  ): Chain[T] => {
    const blockConfig = config.blocks[block];
    const typeConfig = blockConfig.types.find((t) => t.label === type);

    if (!type) {
      const defaultType = blockConfig.types[0];
      if (!defaultType) {
        throw new Error(`No default type found for block ${block}`);
      }

      return {
        type: defaultType.label,
        enabled: false,
        params: defaultType.params.reduce<Record<string, number>>(
          (acc, param) => {
            acc[param.label] = 50;
            return acc;
          },
          {}
        ),
      } as Chain[T];
    }

    if (!typeConfig) {
      throw new Error(`Type ${type} not found in block ${block}`);
    }

    return {
      type: type,
      enabled: true,
      params: typeConfig.params.reduce<Record<string, number>>((acc, param) => {
        acc[param.label] = 50;
        return acc;
      }, {}),
    } as Chain[T];
  };

  return {
    [Blocks.Noisegate]: createTypeParams(
      Blocks.Noisegate,
      types[Blocks.Noisegate]
    ),
    [Blocks.Compressor]: createTypeParams(
      Blocks.Compressor,
      types[Blocks.Compressor]
    ),
    [Blocks.Modulation]: createTypeParams(
      Blocks.Modulation,
      types[Blocks.Modulation]
    ),
    [Blocks.Effect]: createTypeParams(Blocks.Effect, types[Blocks.Effect]),
    [Blocks.Amplifier]: createTypeParams(
      Blocks.Amplifier,
      types[Blocks.Amplifier]
    ),
    [Blocks.Cabinet]: createTypeParams(Blocks.Cabinet, types[Blocks.Cabinet]),
    [Blocks.Eq]: createTypeParams(Blocks.Eq, types[Blocks.Eq]),
    [Blocks.Reverb]: createTypeParams(Blocks.Reverb, types[Blocks.Reverb]),
    [Blocks.Delay]: createTypeParams(Blocks.Delay, types[Blocks.Delay]),
  };
}
