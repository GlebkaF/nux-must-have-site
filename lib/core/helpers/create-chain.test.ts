import { describe, it, expect } from "vitest";

import { createDefaultChain } from "./create-chain";

describe("create-default-chain", () => {
  it("should match snapshot", () => {
    expect(createDefaultChain()).toMatchSnapshot();
  });
});
