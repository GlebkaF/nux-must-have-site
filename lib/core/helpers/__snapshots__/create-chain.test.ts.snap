// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`create-default-chain > should match snapshot 1`] = `
{
  "amplifier": {
    "enabled": true,
    "params": {
      "Bass": 50,
      "Bright": 50,
      "Gain": 50,
      "Master": 50,
      "Middle": 50,
      "Treble": 50,
    },
    "type": "Jazz Clean",
  },
  "cabinet": {
    "enabled": true,
    "params": {
      "HighCut": 50,
      "Level": 50,
      "LowCut": 50,
    },
    "type": "JZ120",
  },
  "compressor": {
    "enabled": true,
    "params": {
      "Clipping": 50,
      "Level": 50,
      "Sustain": 50,
    },
    "type": "K Comp",
  },
  "delay": {
    "enabled": true,
    "params": {
      "Echo": 50,
      "Intensity": 50,
      "Rate": 50,
    },
    "type": "Analog Delay",
  },
  "effect": {
    "enabled": true,
    "params": {
      "Output": 50,
      "Sensitivity": 50,
    },
    "type": "Distortion+",
  },
  "eq": {
    "enabled": true,
    "params": {
      "100": 50,
      "1200": 50,
      "220": 50,
      "2600": 50,
      "500": 50,
      "6400": 50,
    },
    "type": "6-Band",
  },
  "modulation": {
    "enabled": true,
    "params": {
      "Depth": 50,
      "Intensity": 50,
      "Rate": 50,
    },
    "type": "CE-1",
  },
  "noisegate": {
    "enabled": true,
    "params": {
      "Decay": 50,
      "Sensitivity": 50,
    },
    "type": "Noise Gate",
  },
  "reverb": {
    "enabled": true,
    "params": {
      "Decay": 50,
      "Level": 50,
      "Tone": 50,
    },
    "type": "Room",
  },
}
`;
